src/metabase/lib/schema.cljc

```diff
@@ -161,18 +161,34 @@
    [:map
     [:source-card [:ref ::id/card]]]])
 
+(mr/def ::source
+  [:map
+   [:lib/type [:enum :source/metric]]
+   [:id [:ref ::id/metric]]])
+
+(mr/def ::stage.mbql.with-sources
+  [:and
+   [:merge
+    [:ref ::stage.mbql]
+    [:map
+     [:sources [:sequential {:min 1} [:ref ::source]]]]]
+   [:fn
+    {:error/message "A query with :sources cannot have :joins or :fields"}
+    (complement (some-fn :fields :joins))]])
+
 (mr/def ::stage.mbql.with-source
   [:or
    [:ref ::stage.mbql.with-source-table]
-   [:ref ::stage.mbql.with-source-card]])
+   [:ref ::stage.mbql.with-source-card]
+   [:ref ::stage.mbql.with-sources]])
 
 ;;; Schema for an MBQL stage that DOES NOT include `:source-table` -- an MBQL stage that is not the initial stage.
 (mr/def ::stage.mbql.without-source
   [:and
    [:ref ::stage.mbql]
    [:fn
-    {:error/message "Only the initial stage of a query can have a :source-table or :source-card."}
-    (complement (some-fn :source-table :source-card))]])
+    {:error/message "Only the initial stage of a query can have a :source-table, :source-metrics, or :source-card."}
```

Comment:
Unless I'm missing something, this is inconsistent with `:sources` elsewhere. I didn't think `:source-metrics` was a thing?